subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/daemon/DaemonGroovyCompiler.java

```diff
@@ -129,6 +134,10 @@ protected DaemonForkOptions toDaemonForkOptions(GroovyJavaJointCompileSpec spec)
         // This should come from the compiler classpath only
         gradleFilterSpec.disallowPackage("org.gradle.api.internal.tasks.compile");
 
+        gradleFilterSpec.allowClass(ApiCompilerResult.class);
```

Comment:
This shouldn't be necessary, but currently is because the worker API handles return types differently depending on whether you use process isolation or classpath isolation. In the former case, the return value is serialized and deserialized, so the correct class is returned. In the latter case, the result is returned directly, which means it is not an instance of the expected class unless we allow that class to leak through here. Should probably be fixed in the worker API, so that it always serializes/deserializes results, but that seemed out of scope for this change.

Reply:
I think these comments would be good to have in code, too.

Reply:
Done